how-to/customize-workspace/client/src/modules/auth/example/endpoint.ts

```diff
@@ -0,0 +1,205 @@
+import type { ExampleEndpointOptions, ExampleUserRoleMapping } from "./shapes";
+import { getCurrentUser } from "./util";
+
+let logger: Logger;
+let userSessionKey: string;
+let roleMapping: { [key: string]: ExampleUserRoleMapping };
+let definitionData: ModuleDefinition<ExampleEndpointOptions>;
+
+function getRequestOptions(
+	url: string,
+	options: FetchOptions,
+	request: unknown
+): { url: string; options: FetchOptions } {
+	if (options.method === "GET") {
+		if (request !== undefined) {
+			const keys = Object.keys(request);
+			if (keys.length > 0) {
+				const length = keys.length;
+				for (let i = 0; i < length; i++) {
+					url = url.replace(`[${keys[i]}]`, encodeURIComponent(request[keys[i]] as string));
+				}
+			}
+		}
+	} else if (options.method === "POST" && request !== undefined) {
+		options.body = JSON.stringify(request);
+	}
+
+	return { url, options };
+}
+
+function applyCurrentUserToApps(apps: PlatformApp[] = []): PlatformApp[] {
+	const currentUser = getCurrentUser(userSessionKey);
+	if (
+		currentUser === null ||
+		roleMapping === undefined ||
+		roleMapping[currentUser.role] === undefined ||
+		roleMapping[currentUser.role].excludeAppsWithTag === undefined
+	) {
+		return apps;
+	}
+	const excludeTag = roleMapping[currentUser.role].excludeAppsWithTag;
+	const filteredApps: PlatformApp[] = [];
+	for (let i = 0; i < apps.length; i++) {
+		if (Array.isArray(apps[i].tags)) {
+			let include = true;
+			for (let t = 0; t < apps[i].tags.length; t++) {
+				const tag: string = apps[i].tags[t];
+				if (excludeTag.includes(tag)) {
+					include = false;
+					break;
+				}
+			}
+			if (include) {
+				filteredApps.push(apps[i]);
+			}
+		} else {
+			filteredApps.push(apps[i]);
+		}
+	}
+	return filteredApps;
+}
+
+function applyCurrentUserToSettings(settings: CustomSettings): CustomSettings {
+	const currentUser = getCurrentUser(userSessionKey);
+	if (currentUser === null || roleMapping === undefined || roleMapping[currentUser.role] === undefined) {
+		return settings;
+	}
+
+	if (Array.isArray(settings?.endpointProvider?.modules)) {
+		settings.endpointProvider.modules.push({
+			data: definitionData,
+			enabled: definitionData.enabled,
+			id: definitionData.id,
+			description: definitionData.description,
+			icon: definitionData.icon,
+			info: definitionData.info,
+			title: definitionData.title,
+			url: definitionData.url
+		});
+		if (
+			Array.isArray(settings?.endpointProvider?.endpoints) &&
+			Array.isArray(settings?.appProvider?.endpointIds)
+		) {
+			const appEndpoints = settings?.appProvider?.endpointIds;
+			for (let i = 0; i < appEndpoints.length; i++) {
+				if (typeof appEndpoints[i] === "string") {
+					const endpointToUpdate = settings.endpointProvider.endpoints.find(
+						(endpointEntry) => endpointEntry.id === appEndpoints[i] && endpointEntry.type === "fetch"
+					);
+					endpointToUpdate.type = "module";
+					if (endpointToUpdate.type === "module") {
```

Comment:
this is for typescript to be happy with the type because the initial .find looked for a module of type fetch. Added a comment to explain the weirdness.